### ISSUE TITLE
storage: implement proposer-evaluated kv

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#6290]
See #6166. This is the remaining forklift portion after all [sub-issues](https://github.com/cockroachdb/cockroach/labels/proposer-evaluated-kv) have been completed.

**Comment by tschottdorf** [comment]
The work on the correctness-related parts here has now begun, so it's getting interesting.
#10012 adds an explicit "evaluate" step (which as of yet only puts the `BatchRequest` in the proposal, so no proposer-evaluation yet) which does not hold `Replica.mu` (as that would be prohibitively expensive). This is fairly complex because the reproposal and refurbishment code is affected.

Assuming that goes in, there's still a gap before we can actually evaluate proposals early, namely that more reordering protection is lost.

First example: `Put(c)` is proposed together with `Split(b)`. That is, the commands are evaluated concurrently, and then put into Raft in some order. There are no overlapping keys here, so both commands are in-flight concurrently. But of course they shouldn't be; if the Split applies before the write, then the write would hit the wrong Range, which roughly translates to it writing to a part of the Engine that isn't owned by the Range (which has now shrunk), which is wildly illegal (messing up stats on top of enabling various correctness anomalies). So a Split's `EndTransaction` Batch needs to occupy the right half of its split in addition to its regular writes.

Second example: Node1 proposes a command under its lease, but then loses the lease. Later, when another node already holds the lease, the command still applies. That is, we need to keep checking leases downstream of Raft by sending along enough information about the proposer. We currently send `OriginReplica`, but I don't think that's enough. In fact just adding the lease as a whole is not enough. Assume the node power-cycles quickly, so it keeps its lease, but drops its in-memory state. When it comes back, it has forgotten about the inflight command but it must not propose anything overlapping to avoid anomalies. I think when a node comes up, it must not use any lease it acquired before it went down, and I think that's a hole we currently expose (cc @andreimatei who is patching similar holes rooted in not persisting HLC updates).

Third, more subtle example: An aborting Push and a write for the same transaction are inflight simultaneously. They don't overlap; after all, the Push only touches the txn record, the transactional write for that txn does not. Both commands evaluate on a snapshot that doesn't see the other and as it happens, the push applies first. Whoops, the transaction writes a key after it's been aborted. Not yet an anomaly, but I'm sure we can get there somehow. I think transactions always need to at least add themselves in the command queue (ops which mutate the state already should as they write directly to that key).
Similar concerns exist around writes to and checks of the abort cache for a transaction (i.e. any transactional operation and any intent resolutions need to add themselves for those keys as well)

I think there might be oddities regarding transactional reads as well, though I couldn't think of anything in particular.

Going to look at this a bit closer (in particular searching for other writes/state changes that we don't track in the command queue), but appreciate comments.

**Comment by tschottdorf** [comment]
Another issue: during evaluation, we might see inconsistent memory state (for instance, if a command applies and has side effects which lead to more than one critical ReplicaMu section).

**Comment by bdarnell** [comment]
> So a Split's EndTransaction Batch needs to occupy the right half of its split in addition to its regular writes.

That sounds expensive, but I guess it's just the pessimistic version of what we do today (forcing a restart of any writes to the RHS during a split)

>  I think when a node comes up, it must not use any lease it acquired before it went down, and I think that's a hole we currently expose

Agreed, but what's the currently-exposed in-memory state? The tscache is safe due to the low-water mark.

> Whoops, the transaction writes a key after it's been aborted. Not yet an anomaly, but I'm sure we can get there somehow.

This seems fine to me - it's always possible to write an intent on another range concurrent with an abort; why not on the same range? It's a little less efficient since we're going to incur extra round-trips to resolve it, but I don't think there's any correctness risk here. Only the begin, end, and push transaction commands need to be in the command queue for the transaction's key. And I think that as long as those commands are the only ones that touch the abort cache, it's OK not to queue on that key explicitly.

**Comment by tschottdorf** [comment]
> Agreed, but what's the currently-exposed in-memory state? The tscache is safe due to the low-water mark.

I don't think I understand that question. We've dropped the command queue, so we can't guarantee we're going to wait on anything we need to wait for to apply. Btw, the timestamp cache is only safe after #9504 (and, for transfers, #9523), both of which aren't merged yet.

> This seems fine to me - it's always possible to write an intent on another range concurrent with an abort; why not on the same range? It's a little less efficient since we're going to incur extra round-trips to resolve it, but I don't think there's any correctness risk here. Only the begin, end, and push transaction commands need to be in the command queue for the transaction's key. And I think that as long as those commands are the only ones that touch the abort cache, it's OK not to queue on that key explicitly.

I also don't think there's anything inherently incorrect with it but it's surprising, no? Looking at the Raft log, we've aborted a transaction, poisoned the abort cache, and the transaction still went ahead and did a write. There's a read of a key in there that should've waited for an inflight write to it. Remember the recent anomaly we've seen with early transaction aborts? I don't want to make the playing field more complicated unless there's a very specific reason and all implications are well-understood.
For the record, the anomaly I was thinking about here was reading a key we had once written but which had since been deleted, but that can't happen since we populate the AbortCache atomically with the push (and we hit the disk every time), plus the intent is still there. ResolveIntent is also not vulnerable to anything at the time of writing since it actually covers the key with the intent.

I'm going to see what other keys we write by adding some assertions that check that during Raft application, the command queue covers everything that we touch.

**Comment by tschottdorf** [comment]
It was pointed out that one big change with moving the evaluation up is `raftMu` locking (we don't want to hold `raftMu` before proposing to Raft or we lose some of the benefit of the command queue). That's easy to play with without actually moving the evaluation code. I'll see what falls off that tree when I shake it.

**Comment by tschottdorf** [comment]
Ok, I now have a (hacky) asserter that tracks all keys written during a batch and am asserting in `endCmds` that we in fact had all of those in the command queue:

```
W161018 14:24:07.644311 11 storage/replica.go:1249  [r1/1:/M{in-ax}] {/Local/Range/"foo"/Transaction/addrKey:/id:"af9d301b-ce6d-4ee6-a9d4-e8f61fc36e3f" /Min}
F161018 14:24:07.644326 11 storage/replica.go:1251  [r1/1:/M{in-ax}] untracked writes in command queue
```

I'll clean that up and hopefully submit something tomorrow that passes this assertion.

**Comment by bdarnell** [comment]
>> Agreed, but what's the currently-exposed in-memory state? The tscache is safe due to the low-water mark.

> I don't think I understand that question. We've dropped the command queue, so we can't guarantee we're going to wait on anything we need to wait for to apply. Btw, the timestamp cache is only safe after #9504 (and, for transfers, #9523), both of which aren't merged yet.

You said you think we're currently exposing a hole here; I was asking what that hole was. But yeah, now I see how the command queue is a problem too, not just the tscache. 

> I also don't think there's anything inherently incorrect with it but it's surprising, no?

OK, now that I've paged back in how the abort cache works (I thought it only lived on the range with the transaction record) and why we consult it on reads (#2231), I think we do need to consider it here by queuing up all operations for the same transaction.

**Comment by tschottdorf** [comment]
#10084 (mostly) checks what we actually read/write; getting all the dependencies right seems brittle and bad for performance. I'd like to know if the below alternative seems like an obvious non-starter or whether it's worth exploring it as a serious alternative. Happy to hash it out further if it does.

>The proposal's MaxLeaseIndex already makes sure that when we submit command
A to Raft before command B, then they can only apply in that order (mod
refurbishments, which can reorder and cause problems to this arguments,
though I think they can be overcome). So we could imagine a world in which
the command queue wouldn't commonly block commands on top of each other:
>
- a command arrives.
- checks the command queue; finds that it must wait for some prior commands
to
 apply.
- instead of waiting, it evaluates these commands in the correct order and
 applies itself on top of that, entering itself in the command queue but
 immediately proposing to Raft a command whose successful application is
 contingent upon the previous commands applying as planned.
- once the command applies, leave the command queue (as usual)
- if the command does not apply (i.e. it or its dependencies got
reordered),
 leave the command queue and enter it anew (instead of the previous
internal
 refurbishment cycle).
>
The above is fairly half baked and challenging to implement, but so is
tracking every command's implicit writes; the above idea would in principle
remove any latency that isn't Raft or doing actual work, which is a great
benefit.

**Comment by bdarnell** [comment]
I like this as a long-term goal for the latency benefits (the cost of re-applying earlier commands to get a baseline for the new one worries me a bit, but we could mitigate that by caching the prepared batches), but it seems risky to block proposer-evaluated KV on this.

**Comment by bdarnell** [comment]
We have two different ordering mechanisms: the command queue and the lease index. Since the lease index gives a total order to all commands, maybe we don't need a tree of batches but a linear chain: every command is proposed on top of the batch produced by the previous one (and whenever we reset our lease index we rewind to whatever is on disk and start over). Would this free us from having to figure out which commands could interfere with each other for the purposes of proposer-evaluated KV? (this wouldn't let us get rid of the command queue, of course, but it would mean we wouldn't have to expand its role or add additional complications)

**Comment by tschottdorf** [comment]
@bdarnell I replied in https://github.com/cockroachdb/cockroach/pull/10084 since that thread is all about that. TL;DR is that I'm starting to think that #10084 is the right approach to take.
